### PR TITLE
Issue #6: The format of the lock file content can now be specified...

### PR DESCRIPTION
**Pull request by akaihola**
…using an optional keyword argument to LockFile().

This is useful in a container (e.g. Docker) environment.

### CODE REVIEW

**Comment by mgedmin** [comment]
Thank you for the PR!

Have you signed the [Zope Foundation Committer Agreement](http://foundation.zope.org/agreements)?  I'm afraid it's prerequisite to your code getting merged.

It would also be good if you could figure out why the Travis build failed and fix that.

**Comment by akaihola** [comment]
@mgedmin, I just signed and e-mailed the Committer Agreement.

I fixed most of the build errors we saw on Travis, but I'm still trying to find out why PyPy3 [fails with the logging test](https://travis-ci.org/zopefoundation/zc.lockfile/jobs/114531063).

**Comment by akaihola** [comment]
@ssalonen, could we have a review session? I suspect the problem is in my [threaded unit test case](https://github.com/EniramLtd/zc.lockfile/blob/fee0ff6279f78cee107fe0c355d23776af0110d8/src/zc/lockfile/tests.py#L220), not the implementation itself.

**Comment by mgedmin** [comment]
(I do not have the energy to review this, sorry!  Perhaps somebody else can step up?)

**Comment by akaihola** [comment]
Sorry for accidentally including `doc.txt` in one of the commits. If this patch is accepted, the feature branch should probably be squashed into a single commit.

**Comment by akaihola** [comment]
To ease reviewing, here's a copy of the documentation for the new feature:

## Hostname in lock file

In a container environment (e.g. Docker), the PID is typically always
identical even if multiple containers are running under the same operating
system instance.

Clearly, inspecting lock files doesn't then help much in debugging. To identify
the container which created the lock file, we need information about the
container in the lock file. Since Docker uses the container identifier or name
as the hostname, this information can be stored in the lock file in addition to
or instead of the PID.

Use the `content_template` keyword argument to `LockFile` to specify a
custom lock file content format:

    >>> lock = zc.lockfile.LockFile('lock', content_template='{pid};{hostname}')
    >>> lock.close()

If you now inspected the lock file, you would see e.g.:

    $ cat lock
     123;myhostname

**Comment by akaihola** [comment]
@mgedmin, I got a reply from @jimfulton to my Committer Agreement submission:

> You need to supply the name an email of an existing committer who can vouch for you.

Do you think this pull request has a chance to get reviewed even though I don't know any of the existing committers? Do I actually need to become a committer to submit a one-time patch?

**Comment by jimfulton** [comment]
Sorry yes. We'll figure it out. I've been really busy.

Thanks for the pr.

Jim

On Mar 29, 2016 11:27 AM, "Antti Kaihola" <notifications@github.com> wrote:
>
> @mgedmin, I got a reply from @jimfulton to my Committer Agreement
submission:
>>
>> You need to supply the name an email of an existing committer who can
vouch for you.
>
> Do you think this pull request has a chance to get reviewed even though I
don't know any of the existing committers? Do I actually need to become a
committer to submit a one-time patch?

File: src/zc/lockfile/__init__.py [hunk 1]
```diff
@@ -83,15 +128,26 @@ def __init__(self, path):
...
+                values_str = 'pid=UNKNOWN'
             fp.close()
-            if not pid:
-                pid = 'UNKNOWN'
-            logger.exception("Error locking file %s; pid=%s", path, pid)
+            logger.exception("Error locking file %s; %s", path, values_str)
             raise
 
+        # write pid and/or hostname into the lock file
+        content = LockFileContentFormatter().format(content_template)
```
Comment (jimfulton) [inline src/zc/lockfile/__init__.py]: Why not just:

    content = content_template.format(pid=os.getpid(),
                                      hostname=socket.gethostname())

**Comment by jimfulton** [comment]
I like what this PR is trying to do.  The implementation, and thus the PR, is rather large for the feature.

There's way more machinery than needed.  This should be a simple change.  Just change the code to invoke the template with the pid and hostname.

For testing, you seem to be inventing your own mocking machinery.  Just use the mock module. It's included in Python 3 and can be added as a testing dependency for Python 2.  This should make it possible to test your change with wildly less code.

When this is as small a change as it should be, I'll be happy to accept it without you being a Zope contributor.

Thanks again for working on this!

Jim

**Comment by akaihola** [comment]
@jimfulton, thanks for your reply!

`LockFileContentFormatter` is there since I didn't want to import the `socket` module nor call `socket.gethostname()` unless the host name is actually used in the template. I was mainly concerned about compatibility across platforms and exotic environments here.

If you see that always importing `socket` and calling `socket.gethostname()` is not an issue, I'll be happy to simplify that part of the implementation.

`LockFileParserFormatter` adds support for parsing and reporting both pid and hostname in case of a locking failure. We could just include the raw content of the lock file in the exception message instead, if that's ok.

I'll convert tests to use the `mock` module (which I like a lot). I was wary of bringing additional dependencies, but it's good to hear that's not a problem.

**Comment by akaihola** [comment]
@jimfulton, the patch has now been squeezed down a bit. Changes:
- use `unittest.mock` from stdlib or `mock` from PyPI
- only install `mock` for tests if Python < 3.3
- lazy import of `socket` module now done with a simple host name wrapper class instead of `LockFileContentFormatter`
- no parsing of lock file contents if a custom template is provided

**Comment by akaihola** [comment]
@jimfulton, have you had a chance to look at my revised patch?

**Comment by akaihola** [comment]
@jimfulton and @mgedmin, sorry to spam you again, but do you think this still has a chance to get merged? The reason I ask is since we'd prefer to avoid maintaining our own fork at Eniram, so it would be great if this could be included upstream.

All the best from sunny Helsinki,

-Antti

File: src/zc/lockfile/tests.py [hunk 2]
```diff
@@ -94,6 +100,92 @@ def pid_in_lockfile():
     >>> lock.close()
     """
 
+
+def hostname_in_lockfile():
+    r"""
+    # hostname is correctly written into the lock file when it's included in the
+    # lock file content template
```
Comment (mgedmin) [inline src/zc/lockfile/tests.py]: Why the `#` marks?

File: src/zc/lockfile/__init__.py [hunk 3]
```diff
@@ -83,15 +90,24 @@ def __init__(self, path):
...
+                # custom lock file templates
+                logger.exception('Error locking file %s; content: "%s%s"',
+                                 path, content[:40],
+                                 '...' if len(content) > 40 else '')
             raise
 
         self._fp = fp
-        fp.write(" %s\n" % os.getpid())
+        fp.write(" %s\n" % content_template.format(pid=os.getpid(),
+                                                   hostname=LazyHostName()))
```
Comment (mgedmin) [inline src/zc/lockfile/__init__.py]: This is confusing to me.  Why is there a space in front of the pid?

(Why not make it part of the default content_template, i.e. `' {pid}'`?)

File: src/zc/lockfile/tests.py [hunk 4]
```diff
@@ -94,6 +100,92 @@ def pid_in_lockfile():
...
+    >>> import zc.lockfile
+    >>> with patch('socket.gethostname', Mock(return_value='myhostname')):
+    ...     lock = zc.lockfile.LockFile("f.lock", content_template='{hostname}')
+    >>> f = open("f.lock")
+    >>> _ = f.seek(1)
```
Comment (mgedmin) [inline src/zc/lockfile/tests.py]: You don't need the `seek()` here as you're calling `strip()` anyway on the next line.

File: src/zc/lockfile/tests.py [hunk 5]
```diff
@@ -94,6 +100,92 @@ def pid_in_lockfile():
...
+
+    Make sure that locking twice does not overwrite the old hostname:
+
+    >>> lock = zc.lockfile.LockFile("f.lock", content_template='{hostname}')
+    Traceback (most recent call last):
+      ...
+    LockError: Couldn't lock 'f.lock'
+
+    >>> f = open("f.lock")
+    >>> _ = f.seek(1)
```
Comment (mgedmin) [inline src/zc/lockfile/tests.py]: Ditto

File: src/zc/lockfile/tests.py [hunk 6]
```diff
@@ -94,6 +100,92 @@ def pid_in_lockfile():
...
+
+    def exception(self, msg, *args):
+        self.log_entries.append((msg,) + args)
+
+
+class LockFileLogEntryTestCase(unittest.TestCase):
+    """Tests for logging in case of lock failure"""
+    def setUp(self):
+        self.here = os.getcwd()
+        self.tmp = tempfile.mkdtemp()
```
Comment (mgedmin) [inline src/zc/lockfile/tests.py]: Please pass a `prefix='zc.lockfile-test-'` or something like that to `mkdtemp()`, so that stale tempfiles can be traced back to the culprit.

File: src/zc/lockfile/tests.py [hunk 7]
```diff
@@ -94,6 +100,92 @@ def pid_in_lockfile():
...
+        os.chdir(self.tmp)
+
+    def tearDown(self):
+        os.chdir(self.here)
+        setupstack.rmtree(self.tmp)
+
+    def test_log_entry(self):
+        """PID and hostname are parsed and logged from lock file on failure"""
```
Comment (mgedmin) [inline src/zc/lockfile/tests.py]: I'm not a fan of docstrings on test methods, since some testrunners then print the docstring instead of the test class + method name, which makes it harder to find the test.

(I don't remember at the moment whether zope.testrunner does that.)

File: src/zc/lockfile/tests.py [hunk 8]
```diff
@@ -102,4 +194,9 @@ def test_suite():
     suite.addTest(doctest.DocTestSuite(
         setUp=setupstack.setUpDirectory, tearDown=setupstack.tearDown,
         checker=checker))
+    # Add doctests from zc/lockfile/__init__.py
+    suite.addTest(doctest.DocTestSuite(zc.lockfile))
+    # Add unittest test cases from this module
+    suite.addTest(
+        unittest.TestLoader().loadTestsFromModule(sys.modules[__name__]))
```
Comment (mgedmin) [inline src/zc/lockfile/tests.py]: There's a shorter spelling:
```python
suite.addTest(unittest.defaultTestLoader.loadTestsFromName(__name__))
```

File: src/zc/lockfile/tests.py [hunk 9]
```diff
@@ -102,4 +194,9 @@ def test_suite():
     suite.addTest(doctest.DocTestSuite(
         setUp=setupstack.setUpDirectory, tearDown=setupstack.tearDown,
         checker=checker))
+    # Add doctests from zc/lockfile/__init__.py
+    suite.addTest(doctest.DocTestSuite(zc.lockfile))
```
Comment (mgedmin) [inline src/zc/lockfile/tests.py]: This is unrelated to the rest of the changes, yes?  Because I don't see you adding any new doctests to `__init__.py`.  Are there any old ones that were mistakenly exclude?

**Comment by mgedmin** [comment]
Basically LGTM.

File: src/zc/lockfile/__init__.py (hunk 3 above)
Comment (akaihola) [inline src/zc/lockfile/__init__.py]: It was confusing to me why the original implementation insisted on prepending the lock file contents with a space. I wanted to keep the behavior exactly the same if no custom content template is given, and assumed that there's a good reason for the leading space.

But I'll move the space to the default content template instead now.

Comment (mgedmin) [inline src/zc/lockfile/__init__.py]: That may've been premature!

I now see that on Windows the code uses byte-range locking to lock just the 1st byte.  Perhaps locking on Windows prevents other processes from *reading* too?  If that's the case, then my suggestion was seriously wrong!

Comment (akaihola) [inline src/zc/lockfile/__init__.py]: Ok I reverted that and made the tests be more exact in that they skip the first byte, strip any trailing whitespace (just in caseOS specific line endings might become an issue), and compare contents to expected output in the doctest.

Should I squash all the commits into a neat patch or do you prefer to keep the branch history?

**Comment by mgedmin** [comment]
LGTM.

**Comment by akaihola** [comment]
@mgedmin wrote:
> LGTM.

Thanks Marius!

@jimfulton wrote on March 30, 2016:
> When this is as small a change as it should be, I'll be happy to accept it without you being a Zope contributor.

So I guess it's now up to him to accept the pull request?

**Comment by jimfulton** [comment]
Released as 1.2.0. Thanks!
